mesonlib.split_args/quote_arg/join_args

agurtovoy opened the pull request:
Closes #5726. Note that `shlex.quote`/`shlex.split` are still used for building/parsing `MESONINTROSPECT` path (it's [documented behavior](https://mesonbuild.com/Reference-manual.html) *and* `scripts/commandrunner.py` depends on it).

ignatenkobrain commented:
Flake8 detected 9 issues on 28756dfc4e6b99e1a76bf9cce42e7f9929584f81
Visit https://sider.review/gh/repos/19784232/pulls/5835 to review the issues.

<sub>posted by Sider</sub>

dcbaker commented on mesonbuild/mesonlib.py:
```diff
@@ -725,6 +725,81 @@ def has_path_sep(name, sep='/\\'):
+                    arg += (num_backslashes // 2) * '\\'
+                    num_quotes += 1
+                    in_quotes = not in_quotes
+                elif c in _whitespace and not in_quotes and (len(arg) or num_quotes):
```
`arg or num_quotes`

dcbaker commented on mesonbuild/mesonlib.py:
```diff
@@ -725,6 +725,81 @@ def has_path_sep(name, sep='/\\'):
+
+                num_backslashes = 0
+
+        if len(arg) or num_quotes:
```
same here

dcbaker commented on mesonbuild/mesonlib.py:
```diff
@@ -725,6 +725,81 @@ def has_path_sep(name, sep='/\\'):
+
+if is_windows():
+    # See https://docs.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments
+    # # and https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
```
a tl;dr would be nice, something like:
```
shlex.split isn't safe to use with Windows CMD because of \\, these functions are.
```

ignatenkobrain commented on mesonbuild/mesonlib.py:
```diff
@@ -725,6 +725,83 @@ def has_path_sep(name, sep='/\\'):
 
+
+if is_windows():
+    # shlex.split is not suitable for splitting command line on Window (https://bugs.python.org/issue1724822); 
```
__[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5835#issue-4472770))

<sub>posted by Sider</sub>

jpakkane commented:
I reran the tests and the build failure was not sporadic. Could you rebase this against current master to see if this is fixed in current master?

agurtovoy commented:
@jpakkane Looks like it didn't help. `msys2 gccx64ninja` fails for everyone with the same error (see below), but `msys2 clangx64ninja` passes in other PRs. I'll take a look at it on Tue.

```
internal error in mingw32_gt_pch_use_address, at config/i386/host-mingw32.c:186: MapViewOfFileEx: Attempt to access invalid address. 
```

agurtovoy commented:
@jpakkane `msys2 clangx64ninja` is now passing; it turns out `llvm-config` can return an option with newlines in it, and my implementation of `split_args` wasn't prepared for that. Fixed it, added more unit tests.

agurtovoy commented:
@jpakkane This is ready to be merged.

jon-turney commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -44,11 +43,14 @@
 if mesonlib.is_windows():
+    # FIXME: can't use quote_arg on Windows just yet; there are a number of existing workarounds
+    # throughout the codebase that cumulatively make the current code work (see, e.g. Backend.escape_extra_args
+    # and NinjaBuildElement.write below) and need to be properly untangled before attempting this
```
I did some (most?) of that untangling as part of #5245, maybe that's something I should look at extracting from that PR.
